src/realm.js

```diff
@@ -1295,16 +1310,49 @@ export class Realm {
 
   // Record the current value of binding in this.modifiedBindings unless
   // there is already an entry for binding.
-  recordModifiedBinding(binding: Binding): Binding {
+  recordModifiedBinding(binding: Binding, value?: Value): Binding {
     if (binding.environment.isReadOnly) {
       // This only happens during speculative execution and is reported elsewhere
       throw new FatalError("Trying to modify a binding in read-only realm");
     }
-    if (this.modifiedBindings !== undefined && !this.modifiedBindings.has(binding))
+
+    const isEnvSafe = root => {
+      let context = this.getRunningContext();
+      let { lexicalEnvironment: env, function: func } = context;
+
+      invariant(func instanceof FunctionValue);
+      if (func === root.$FunctionObject) {
+        return true;
+      }
+      if (this.createdObjectsTrackedForLeaks !== undefined && !this.createdObjectsTrackedForLeaks.has(func)) {
+        return false;
+      }
+      env = env.parent;
+      while (env) {
+        if (env.environmentRecord === root) {
+          return true;
+        }
+        env = env.parent;
+      }
+      return false;
+    };
+
+    if (this.modifiedBindings !== undefined && !this.modifiedBindings.has(binding)) {
+      if (value !== undefined && this.isInPureScope() && this.reportSideEffectCallback !== undefined) {
+        let env = binding.environment;
+
+        if (
+          !(env instanceof FunctionEnvironmentRecord) ||
```

Comment:
The binding can be other environment records that are not safe. Like with, block scoped variables (let/const)...

Reply:
Will add other bindings, do you know which ones you were thinking of?

Reply:
I think you can add an invariant here that `env instanceof DeclarativeEnvironmentRecord`. If it's one of the other environments, like the ones that write to objects (`with` and global), it should not get recorded here. It'll be recorded as a modified property. That's probably a good comment to add.

However, it might not be a function environment record.

If it is a let/const, it should be a DeclarativeEnvironmentRecord. To figure out what function it belongs to you have to traverse its parent until you hit a function or global. That's the function that was used to create the environment record. That function is what you use to look up in the map.